File: julia/monte_carlo_parallel.jl
include("xoshiro256.jl")
using Base.Threads

const PI_THEORETICAL = 3.141592653589793238462643383279502884197
const SEED_MULTIPLIER = 0x9E3779B97F4A7C15

function calculate_pi(iterations::Int64, num_threads::Int)
    iterations_per_thread = iterations ÷ num_threads
    base_seed = UInt64(12345)
    
    results = zeros(Int64, num_threads)
    
    @threads for thread_id in 1:num_threads
        # 推奨方式: シードの加算幅を大きくする
        thread_seed = base_seed + (UInt64(thread_id - 1) * SEED_MULTIPLIER)
        rng = Xoshiro256(thread_seed)
        
        inside_circle = 0
        
        for i in 1:iterations_per_thread
            x = next_double(rng)
            y = next_double(rng)
            
            if x * x + y * y <= 1.0
                inside_circle += 1
            end
        end
        
        results[thread_id] = inside_circle
    end
    
    total_inside = sum(results)
    pi_estimate = 4.0 * total_inside / iterations
    error = abs(pi_estimate - PI_THEORETICAL)
    
    return pi_estimate, error
end

function main()
    iterations = parse(Int64, length(ARGS) > 0 ? ARGS[1] : "100000000")
    num_threads = Threads.nthreads()
    
    start = time()
    pi_estimate, error = calculate_pi(iterations, num_threads)
    elapsed = time() - start
    
    elapsed_ms = elapsed * 1000.0
    
    println("{")
    println("  \"language\": \"Julia\",")
    println("  \"variant\": \"standard\",")
    println("  \"version\": \"$(VERSION)\",")
    println("  \"mode\": \"parallel\",")
    println("  \"iterations\": $iterations,")
    println("  \"pi_estimate\": $pi_estimate,")
    println("  \"error\": $error,")
    println("  \"time_ms\": $elapsed_ms,")
    println("  \"memory_mb\": 0.0,")
    println("  \"cache_misses\": 0,")
    println("  \"lines_of_code\": 0,")
    println("  \"compiler_flags\": \"N/A (JIT compiled)\",")
    println("  \"cpu_model\": \"N/A\",")
    println("  \"cpu_cores\": $(Sys.CPU_THREADS),")
    println("  \"thread_count\": $num_threads,")
    println("  \"os\": \"$(Sys.KERNEL)\",")
    println("  \"os_version\": \"N/A\",")
    println("  \"compiler\": \"julia\",")
    println("  \"simd_detected\": false,")
    println("  \"simd_instructions\": []")
    println("}")
end

main()

File: go/monte_carlo_parallel.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"runtime"
	"sync"
	"time"
)

const PI_THEORETICAL = 3.141592653589793238462643383279502884197
const SEED_MULTIPLIER = 0x9E3779B97F4A7C15

func calculatePi(iterations uint64, numGoroutines int) (float64, float64) {
	iterationsPerGoroutine := iterations / uint64(numGoroutines)
	baseSeed := uint64(12345)

	var wg sync.WaitGroup
	results := make([]uint64, numGoroutines)

	for i := 0; i < numGoroutines; i++ {
		wg.Add(1)
		go func(threadID int) {
			defer wg.Done()

			threadSeed := baseSeed + (uint64(threadID) * SEED_MULTIPLIER)
			rng := NewXoshiro256(threadSeed)

			insideCircle := uint64(0)
			for j := uint64(0); j < iterationsPerGoroutine; j++ {
				x := rng.NextDouble()
				y := rng.NextDouble()

				if x*x+y*y <= 1.0 {
					insideCircle++
				}
			}

			results[threadID] = insideCircle
		}(i)
	}

	wg.Wait()

	totalInside := uint64(0)
	for _, count := range results {
		totalInside += count
	}

	piEstimate := 4.0 * float64(totalInside) / float64(iterations)
	error := piEstimate - PI_THEORETICAL
	if error < 0 {
		error = -error
	}

	return piEstimate, error
}

func main() {
	iterations := uint64(100000000)
	numGoroutines := runtime.NumCPU()

	start := time.Now()
	piEstimate, error := calculatePi(iterations, numGoroutines)
	elapsed := time.Since(start)

	elapsedMs := float64(elapsed.Nanoseconds()) / 1e6

	result := map[string]interface{}{
		"language":      "Go",
		"variant":       "standard",
		"version":       "1.21",
		"mode":          "parallel",
		"iterations":    iterations,
		"pi_estimate":   piEstimate,
		"error":         error,
		"time_ms":       elapsedMs,
		"memory_mb":     0.0,
		"cache_misses": 0,
		"lines_of_code": 0,
		"compiler_flags": "go build -ldflags=\"-s -w\"",
		"cpu_model":     "N/A",
		"cpu_cores":      numGoroutines,
		"thread_count":   numGoroutines,
		"os":            runtime.GOOS,
		"os_version":    "N/A",
		"compiler":      "go",
		"simd_detected": false,
		"simd_instructions": []string{},
	}

	jsonData, _ := json.MarshalIndent(result, "", "  ")
	fmt.Println(string(jsonData))
	os.Exit(0)
}

